adage/genes/views.py

```diff
@@ -1,3 +1,57 @@
-from django.shortcuts import render
+from rest_framework.viewsets import ReadOnlyModelViewSet
+from .models import Gene
+from .serializers import GeneSerializer
 
-# Create your views here.
+
+class GeneViewSet(ReadOnlyModelViewSet):
+    """Genes viewset"""
+
+    http_method_names = ['get']
+    serializer_class = GeneSerializer
+
+    def get_queryset(self):
+        queryset = Gene.objects.all()
+
+        # 'search' parameter, which does full text search based on the
+        # following four fields in Gene model:
+        # - "standard_name": highest priority (A: 1.0);
+        # - "systematic_name": second highest priority (B: 0.4);
+        # - "description" and "crossref__xrid": lowest priority (C: 0.2).
+        search_str = self.request.query_params.get('search', None)
+        if search_str is not None:
+            from django.contrib.postgres.search import (
+                SearchQuery, SearchRank, SearchVector
+            )
+
+            standard_vector = SearchVector('standard_name', weight='A')
+            systematic_vector = SearchVector('systematic_name', weight='B')
+            other_vector = SearchVector(
+                'description', 'crossref__xrid', weight='C'
+            )
+
+            vectors = standard_vector + systematic_vector + other_vector
+            query = SearchQuery(search_str)
+            queryset = queryset.annotate(
+                rank=SearchRank(vectors, query)
+            ).filter(rank__gte=0.1
+            ).order_by('-rank', 'standard_name')
+
+        # 'autocomplete' parameter, which does trigram search based on the
+        # following three fields in Gene model:
+        # - "standard_name" (multiplied by 2.0 to give it a higher priority)
+        # - "systematic_name"
+        # - "description"
+        similarity_str = self.request.query_params.get('autocomplete', None)
+        if similarity_str is not None:
+            from django.contrib.postgres.search import TrigramSimilarity
+
+            queryset = queryset.annotate(
+                similarity=(
+                    TrigramSimilarity('standard_name', similarity_str) * 2.0 +
+                    TrigramSimilarity('systematic_name', similarity_str) +
+                    TrigramSimilarity('description', similarity_str)
```

Comment:
I am not sure that the description should be included in the trigram search. Since this is for autocomplete, shouldn't it only be indexing the terms we want to fill in on the field that's being completed? Indexing anything other than what belongs in the field will just create noise in the autocomplete list, I think.

Reply:
Hmm. It feels wrong that this index is happening _during_ the query. Shouldn't this be precomputed somehow? We may not see any issues with a small test index, but attempting to scale this up will potentially cause problems.

As I mentioned above, I need to read more about Postgres full-text searching, so maybe I'm missing something.

Reply:
The original autocomplete implementation in Elasticsearch was based on Gene's `wall_of_name`, which includes `description` field. @cgreene, do you think it makes more sense to exclude gene description in autocomplete search? It does make some sense to me.

About the search indexing, It's possible to build it beforehand by modifying `models.py` (just like adding  other regular indexes). I didn't do it for a couple of reasons:
- It is Postgres-specific. We have to use if/else to handle it in another database such as `SQLite`.
- The performance seems acceptable even if there is no indexing. 

We have 5,739 genes now in the database. I agree that this could be an issue when we scale up in the future.

Reply:
I am very impressed if the test database contains 5,739 entries and the search is performing as well as it is by regenerating the index for each call.

There are two aspects to scalability at issue here that I can see. One is scaling on the number of genes in the database. The other is scaling on the number of concurrent users hitting the site. 

We can test for scaling on number of genes by setting up a timed test and running it with 10, 100, 500, 1000, and all 5,739 genes. I would be surprised if the time to execute each test scales linearly.

There are tools for simulating lots of API calls also. Getting that set up properly requires having full end-to-end tests.

Reply:
I've opened #10 to document my performance concerns here.